### PR TITLE
feat: proposal for Section component onToggle prop

### PR DESCRIPTION
**Pull request by aesqe**
[SPB-3213]

### Description

Proposal for a new prop for the `Section` component, please see the [`.md` file in the proposals folder](https://github.com/toptal/picasso/blob/86215402166984733c5ec0b79bbaeff7d10a6d1a/proposals/section-component-ontoggle-prop.md)

### Development checks

- [ ] Add changeset according to [guidelines](https://github.com/toptal/picasso/blob/master/docs/contribution/changeset-guidelines.md) (if needed)
- [ ] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [ ] Annotate all `props` in component with documentation
- [ ] Create `examples` for component
- [ ] Ensure that deployed demo has expected results and good examples
- [ ] Ensure that tests pass by running `yarn test`
- [ ] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [ ] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).
- [ ] Self reviewed
- [ ] Covered with tests

**Breaking change**

- [ ] codemod is created and showcased in the changeset
- [ ] test alpha package of Picasso in StaffPortal

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[SPB-3213]: https://toptal-core.atlassian.net/browse/SPB-3213?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: proposals/section-component-ontoggle-prop.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+```jsx
+import React, { useState } from 'react'
+import { Link, Section } from '@toptal/picasso'
+
+const ExampleComponent = () => {
+  const [collapsed, setCollapsed] = useState(false)
+  const title = collapsed ? 'Toptal' : <Link href='https://toptal.com'>Toptal</Link>
+
+  return (
+    <Section collapsible onToggle={setCollapsed} title={title}>
```
Comment (OndrejTuma) [inline proposals/section-component-ontoggle-prop.md:43]: I'm wondering if this solution would fit you better:
```suggestion
    <Section collapsible onToggle={setCollapsed} title={collapsed => collapsed ? 'Toptal' : <Link href='https://toptal.com'>Toptal</Link>}>
```
You wouldn't need app logic anymore

Comment (TomasSlama) [reply proposals/section-component-ontoggle-prop.md:43]: @OndrejTuma `onToggle` might be used for anything else 🤔. I like it more.

Comment (OndrejTuma) [reply proposals/section-component-ontoggle-prop.md:43]: I agree. It can be added as well. I was thinking about solving this particular issue 🙂

**Review by OleksandrNechai — APPROVED** [review]
I think we made a mistake some time ago by making this component control its collapsed state. We should have done it controlled.
But given we already have uncontrolled in place, I think this new prop makes sense.

File: proposals/section-component-ontoggle-prop.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+(...)
+export const Section = ((...) props, ref ) {
+  const {
+(...)
+    onToggle, // Added
+    ...rest
+  } = props
+(...)
+  const toggleCollapse = () => {
+    onToggle?.(!collapsed) // Added
```
Comment (dmaklygin) [inline proposals/section-component-ontoggle-prop.md:26]: the only question I have is:
we have an animation and the Section state might have two values: when animation is started, when animation is finally finished.
you call `onToggle` when the animation has just started (frankly, even when it is not started yet).
Do we want to call onToggle when animation is started or when is already finished?

Comment (tamatomas) [reply proposals/section-component-ontoggle-prop.md:26]: Imagine performing a huge calculation on animation started and making the animation look creepy/slow, but it depends, can it be provided as a parameter to `onToggle`?

Comment (aesqe) [reply proposals/section-component-ontoggle-prop.md:26]: @dmaklygin I guess it makes sense to do as little work as possible 🤷 So, that would mean delaying the callback by the animation duration, which is 300ms, right? Is there a standardized way to do it in Picasso, or just using `setTimeout`?

Comment (OleksandrNechai) [reply proposals/section-component-ontoggle-prop.md:26]: I think `onToggle` is something that should fire, well, on a toggle. Not on animation end or start. If that is needed, we will need extra callbacks. Hence 👎 to delaying callback.

**Review by TomasSlama — APPROVED** [review]

**Review by OndrejTuma — APPROVED** [review]

**Comment by aesqe** [comment]
@OleksandrNechai so what are the next proper steps? Does this PR get merged and then the proposal file is removed in the implementation PR? Thanks! 🙇

**Review by denieler — COMMENTED** [review]
I think if we want this component to be collapsible and to react on collapse - we need to make it controlled:

```
const [collapsed, setCollapsed] = useState(false)
const onCollapse = useCallback(collapsed => {}, [])

<Section collapsed={collapsed} onCollapse={onCollapse} />
```

Other than that I think Section component shouldn't be responsible for collapsing the content. We can have a separate component `Collapse`, as MUI does - https://mui.com/material-ui/react-list/#nested-list

**Review by dmaklygin — APPROVED** [review]

**Comment by OleksandrNechai** [comment]
> @OleksandrNechai so what are the next proper steps? Does this PR get merged and then the proposal file is removed in the implementation PR? Thanks! 🙇

@aesqe Daniel suggests that we turn components to the controlled one. So I guess next steps for yo would be to rework this proposal to `controlled` component and reiterate with approvals.

**Comment by OleksandrNechai** [comment]
Since we prefer approaching this problem by making Section a controlled component, this proposal is rejected. New proposals are welcome (i.e. proposal to make Section controlled component)  🙇

**Comment by aesqe** [comment]
@OleksandrNechai I agree that making it a controlled component would be the best approach 👍 As it turns out, we actually don't need the functionality right now because the title was supposed to be a link the whole time and I made an assumption based on the design 😞 However, it would be a nice opportunity to make a contribution to Picasso with this change, so I'll probably create a new proposal soon 🙇 One other thing: how come you have decided to use PRs for RFCs and not GitHub Discussions? Not that it's problematic, just wondering 😄 Thanks!

**Comment by OleksandrNechai** [comment]
@aesqe hey, man!
We have not yet decided, we are just experimenting. The idea is that we want to get a document as a result, not just a discussion. And we can reference that doc later in PR and keep it for the history.

**Pull request closed without merging**